### PR TITLE
KafkaMirrorMaker2: Support MM2 connectors pause/resume and status

### PR DESCRIPTION
**Pull request by ajborley**
### Type of change

Enhancement / new feature

### Description

Based on the enhancement to the KafkaConnector operator and spec in #2363, this PR uses the pause property in 
KafkaMirrorMaker2.spec.mirror.[sourceConnector|checkpointConnector|heartbeatConnector] to control pausing/resuming MirrorMaker 2.0 connectors running in a KafkaMirrorMaker2 deployment (which is a Kafka Connect cluster). 

The PR also adds a connectors property to the KafkaMirrorMaker2.status to expose a list of all the configured MirrorMaker 2.0 connectors and their statuses. Each status is a JSON result of the
`GET /connectors/<MM2ConnectorName>/status` endpoint. With this change a typical status for a KafkaMirrorMaker2 CR with a single mirror using the checkpointConnector and the sourceConnector looks like like this:

```
status:
  conditions:
  - lastTransitionTime: "2020-02-02T23:51:15.943Z"
    status: "True"
    type: Ready
  connectors:
  - connector:
      state: RUNNING
      worker_id: 10.254.22.65:8083
    name: my-cluster-source->my-cluster-target.MirrorCheckpointConnector
    tasks:
    - id: 0
      state: RUNNING
      worker_id: 10.254.22.65:8083
    type: source
  - connector:
      state: PAUSED
      worker_id: 10.254.22.65:8083
    name: my-cluster-source->my-cluster-target.MirrorSourceConnector
    tasks:
    - id: 0
      state: RUNNING
      worker_id: 10.254.22.65:8083
    type: source
  observedGeneration: 2
  url: http://my-cluster-mirrormaker2-api.mirrormaker2-cluster-test.svc:8083

```

As with the KafkaConnector PR, note that the REST API calls to /pause and /resume are asynchronous and the operator doesn't poll waiting for all tasks to stop. This means the status example above (with a PAUSED connector and a RUNNING task) would exist until the next reconciliation (when the status would be updated to show a PAUSED task--assuming the task had indeed stopped) even though the task was actually paused much sooner than that.

Signed-off-by: Andrew Borley <borley@uk.ibm.com>

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/status/ConnectorStatus.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ */
+@Buildable(
+        editableEnabled = false,
+        generateBuilderPackage = false,
+        builderPackage = "io.fabric8.kubernetes.api.builder"
+)
+@JsonInclude(JsonInclude.Include.NON_DEFAULT)
+@JsonPropertyOrder({ "name", "status" })
+@EqualsAndHashCode
+public class ConnectorStatus implements Serializable, UnknownPropertyPreserving {
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/status/ConnectorStatus.java]: How does this fit with the `KafkaConnectorStatus` class? It seems there is some overlap, I wonder if we can reuse it or something.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java [hunk 2]
```diff
@@ -358,6 +362,47 @@ private static void addClusterToMirrorMaker2ConnectorConfig(Map<String, Object>
...
+                                            apiClient.status(host, KafkaConnectCluster.REST_API_PORT,
+                                                    connectorName));
+                        } else if ("PAUSED".equals(state) && !shouldPause) {
+                            log.debug("{}: Resuming connector {}", reconciliation, connectorName);
+                            return apiClient.resume(host, KafkaConnectCluster.REST_API_PORT,
+                                    connectorName)
+                                    .compose(ignored ->
+                                            apiClient.status(host, KafkaConnectCluster.REST_API_PORT,
+                                                    connectorName));
+
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java]: This looks to be more or less identical to the code we use for `KafkaConnectors`. Can we factor out a common method?

File: api/src/main/java/io/strimzi/api/kafka/model/status/ConnectorStatus.java (hunk 1 above)
Comment (ajborley) [reply api/src/main/java/io/strimzi/api/kafka/model/status/ConnectorStatus.java]: I have updated the CrdGenerator to support a list of java.util.Map object, which makes the KafkaMirrorMaker2Status class more closely line up with the KafkaConnectorStatus and means this class (`ConnectorStatus`) can be removed from this PR.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java (hunk 2 above)
Comment (ajborley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java]: Agreed - the latest commit factors out the common code.

**Comment by scholzj** [comment]
Thanks for the PR @ajborley

**Pull request merged**